src/network.c

```diff
@@ -2204,55 +2204,54 @@ static int sockent_add(sockent_t *se) /* {{{ */
 static void *dispatch_thread(void __attribute__((unused)) * arg) /* {{{ */
 {
   while (42) {
-    receive_list_entry_t *ent;
-    sockent_t *se;
-
     /* Lock and wait for more data to come in */
     pthread_mutex_lock(&receive_list_lock);
     while ((listen_loop == 0) && (receive_list_head == NULL))
       pthread_cond_wait(&receive_list_cond, &receive_list_lock);
 
-    /* Remove the head entry and unlock */
-    ent = receive_list_head;
-    if (ent != NULL)
-      receive_list_head = ent->next;
-    receive_list_length--;
-    pthread_mutex_unlock(&receive_list_lock);
-
     /* Check whether we are supposed to exit. We do NOT check `listen_loop'
      * because we dispatch all missing packets before shutting down. */
-    if (ent == NULL)
+    if (receive_list_head == NULL)
       break;
 
-    /* Look for the correct `sockent_t' */
-    se = listen_sockets;
-    while (se != NULL) {
-      size_t i;
+    /* Remove the list and unlock */
+    receive_list_entry_t *private_list_head = receive_list_head;
+    receive_list_head = NULL;
+    receive_list_tail = NULL;
+    receive_list_length = 0;
+    pthread_mutex_unlock(&receive_list_lock);
 
-      for (i = 0; i < se->data.server.fd_num; i++)
-        if (se->data.server.fd[i] == ent->fd)
-          break;
+    while (private_list_head != NULL) {
+      receive_list_entry_t *ent = private_list_head;
 
-      if (i < se->data.server.fd_num)
-        break;
+      /* Look for the correct `sockent_t' */
+      sockent_t *se = listen_sockets;
+      while (se != NULL) {
+        size_t i;
 
-      se = se->next;
-    }
+        for (i = 0; i < se->data.server.fd_num; i++)
```

Comment:
Please declare `i` inside the for loop:

```c
for (size_t i = 0; …
//   ^^^^^^
```

Reply:
I think we shouldn't do that:
```
while (se != NULL) {
  size_t i;

  for (i = 0; i < se->data.server.fd_num; i++)
    if (se->data.server.fd[i] == ent->fd)
      break;

  if (i < se->data.server.fd_num)
    break;

  se = se->next;
}
```
It's the same code that main but indented as now run inside the loop that process whole queue.